### ISSUE TITLE
Using askOption with testSpec (or any IO)

### ISSUE THREAD

**Issue by dgaw** [feuerbach/tasty#228]
Is there a way of using `askOption` with `testSpec` (or IO in general) that doesn't involve `unsafePerformIO`?

**Comment by feuerbach** [comment]
The trick is to call `askOption` *outside* of `testSpec`, not inside it.

    askOption $ \my_option_value -> testSpec $
      {- do something with my_option_value -}`

**Comment by dgaw** [comment]
Thanks for the quick reply @feuerbach. 

I'm not sure how I can apply your suggestion since testSpec produces an `IO TestTree` rather than a plain `TestTree` that `askOption` expects as a result of its lambda argument.

testSpec has a signature:
`testSpec :: TestName -> Spec -> IO TestTree`

askOption's has this:
`IsOption v => (v -> TestTree) -> TestTree`

Perhaps it would make sense to have a version of askOption like so:
`(IsOption v, Monad m) => (v -> m TestTree) -> m TestTree`

Or is there a simpler solution?

**Comment by mitchellwrosen** [comment]
@feuerbach `testSpec` returns an IO value because in hspec, test trees are generated in IO (however I suspect the vast majority of test suites have a static test tree).

**Comment by mitchellwrosen** [comment]
Here is the type: https://hackage.haskell.org/package/hspec-core-2.5.6/docs/Test-Hspec-Core-Spec.html#t:SpecM

**Comment by feuerbach** [comment]
Right. So @dgaw, I don't know if I decide to support IO-generated tests, but even if I do, it won't be any time soon. I'm afraid for now you're stuck with one of the workarounds.

**Comment by dgaw** [comment]
@feuerbach No worries, I can live with that :) So far I haven't noticed any adverse effects of using `unsafePerformIO` on my particular test suite.

**Comment by dcastro** [comment]
I've been using `unsafePeformIO` inside `askOption`, as suggested here, and have just found a scenario where this can go wrong, so I'm posting it here as a word of caution for future readers.

In my case, I had a boolean switch option. We read the option using `askOption` and:
* if it's not set, we do some IO, and build an inner TestTree according to the result of the IO.
* if it's set, it is absolutely important that we don't run any IO or it'll crash, and we build an empty `testGroup` instead.

```hs
newtype NoRunNetworkOpt = NoRunNetworkOpt Switch
  deriving stock (Show, Eq)

data Switch = Unset | Set
  deriving stock (Show, Eq)

main :: IO ()
main =
  defaultMainWithIngredients
    (includingOptions [Tasty.Option $ Proxy @NoRunNetworkOpt] : defaultIngredients)
    tree

tree :: TestTree
tree =
  askOption @NoRunNetworkOpt $ \case
    NoRunNetworkOpt Unset ->
      testGroup "group1"
        if unsafePerformIO (putStrLn "unsafe things go here" >> pure True)
          then [...]
          else []
    NoRunNetworkOpt Set ->
      -- Can't run IO here!
      testGroup "group1" []
```

The problem is that, even when the switch is set from the command line, the IO bit will still run.

That's because, before running the tests, tasty will first traverse the TestTree **with an empty `OptionSet`**, to figure out which options are needed to run it.

See: 
> Test.Tasty.CmdLine.defaultMainWithIngredients ->
> Test.Tasty.CmdLine.parseOptions ->
> Test.Tasty.CmdLine.suiteOptionParser ->
> Test.Tasty.Ingredients.suiteOptions ->
> Test.Tasty.Core.treeOptions

```hs
treeOptions :: TestTree -> [OptionDescription]
treeOptions =

  Prelude.concat .
  Map.elems .

  foldTestTree
    trivialFold { foldSingle = \_ _ -> getTestOptions }
    mempty -- <-- Empty `OptionSet`
```

In other words, if you have `IO` that depends on the value of an option, it's only safe to run it inside the test implementation (e.g. inside `testCase`), and never outside.
